File: src/irrgarten/Labyrinth.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;
import java.util.ArrayList;

/**
 *
 * @author manuel
 */
public class Labyrinth {
    private static final char BLOCK_CHAR ='X';
    private static final char EMPTY_CHAR ='-';
    private static final char MONSTER_CHAR ='M';
    private static final char COMBAT_CHAR ='C';
    private static final char EXIT_CHAR ='E';
    private static final int ROW=0;
    private static final int COL=1;
    private int nRows;
    private int nCols;
    private int exitRow;
    private int exitCol;
    private char [][] labyrinth;
    private Monster [][] monsters;
    private Player [][] players;
    
    Labyrinth (int nCols, int nRows, int exitRow, int exitCol){
        this.nRows = nRows;
        this.nCols = nCols;
        this.exitRow = exitRow;
        this.exitCol = exitCol;

        labyrinth = new char[nRows][nCols];
        monsters = new Monster[nRows][nCols];
        players = new Player[nRows][nCols];
        
        for (int i=0; i< nCols;i++){
            for (int j = 0; j < nRows; j++){
                this.labyrinth[i][j]=EMPTY_CHAR;
                this.monsters[i][j] = null;
                this.players[i][j] = null;
            }
        }
        this.labyrinth[exitRow][exitCol]=EXIT_CHAR;
    } 
      

    void spreadPlayers(ArrayList<Player> jugadores){
        for (Player player:  jugadores) {
            int [] pos = new int [2];
            pos = randomEmptyPos();
            putPlayer2D(-1, -1, pos[ROW], pos[COL], player);
        }

    }

    public boolean haveAWinner(){
        return !(players[exitRow][exitCol]== null);
    }
    
    @Override
    public String toString() {
        String cadena="", marco="";

        for (int i=0; i<nCols+2;i++){
            marco +=" * ";
        }
        marco += "\n";
        cadena += marco;

        for (int i=0; i < nRows; i++){
            cadena += " * ";
            for (int j=0; j < nCols; j++){
                cadena += " " + this.labyrinth[i][j] + " ";
            }
            cadena += " *\n";
        }
        cadena += marco;

        return cadena;
    }
    
    public void addMonster (int row, int col, Monster monster){
        if (this.posOK(row, col) && labyrinth[row][col]==EMPTY_CHAR){
            this.labyrinth[row][col]= MONSTER_CHAR;
            this.monsters[row][col]= monster;
        }
    }
    
    public Monster putPlayer(Directions direction, Player player){
        int[] newPos = new int[2];
        Integer oldRow = player.getRow();
        Integer oldCol = player.getCol();

        newPos = this.dir2Pos(oldRow,oldCol, direction);
        return this.putPlayer2D(oldRow,oldCol , newPos[ROW], newPos[COL], player);  
    }
    
    public void addBlock(Orientation orientation, int startRow, int startCol, int length){
        int incRow, incCol;
        if (orientation == Orientation.VERTICAL){
            incRow = 1;
            incCol = 0;
        }
        else {
            incRow = 0;
            incCol = 1;
        }

        int row = startRow;
        int col = startCol;

        while (posOK(row, col) && emptyPos(row, col) && length > 0){
            labyrinth[row][col] = BLOCK_CHAR;
            length -= 1;
            row += incRow;
            col += incCol;
        }
    }   
    
    //Devuelve array de direcciones validas 

    public ArrayList<Directions> validMoves(Integer row,Integer col){
        ArrayList<Directions>output = new ArrayList<>();

        if (this.canStepOn(row +1, col)){
            output.add(Directions.DOWN);
        }
        if (this.canStepOn(row-1,col)) {
            output.add(Directions.UP);
        }
        if (this.canStepOn(row,col+1)) {
            output.add(Directions.RIGHT);
        }
        if (this.canStepOn(row,col-1)) {
            output.add(Directions.LEFT);
        }
        return output;
    }
    
    private boolean posOK(int row, int col){
         return (0 <= row && row < nRows && 0 <= col && col < nCols);
    }
    
    private boolean emptyPos(int row, int col){
        return (this.labyrinth[row][col] == EMPTY_CHAR);
          
    }

    private boolean monsterPos(int row, int col){
        return (this.labyrinth[row][col]==MONSTER_CHAR);

    }

    private boolean exitPos(int row, int col){
        return (this.labyrinth[row][col]==EXIT_CHAR);
    }   

    private boolean combatPos(int row, int col){
        return (this.labyrinth[row][col]==COMBAT_CHAR);

    }

    private boolean canStepOn(int row, int col){
        if (posOK(row, col)){
            if(this.emptyPos(row, col) || this.monsterPos(row, col) || this.exitPos(row, col)){
                return true;
            }
            else return false;
        }     
        else 
            return false;
    }

    private void updateOldPos(int row, int col){
        if (posOK(row, col)){
            if (this.combatPos(row, col)){
                this.labyrinth[row][col]= MONSTER_CHAR;
            }
            else this.labyrinth[row][col]= EMPTY_CHAR;
        }
    }

    private int[] dir2Pos(int row, int col, Directions direction){
        int [] newPos = new int[2];
        if (direction == Directions.UP){
            newPos[0]=row-1;
            newPos[1]=col;
        }
        if (direction == Directions.DOWN){
            newPos[0]=row+1;
            newPos[1]=col;
        }
        if (direction == Directions.LEFT){
            newPos[0]=row;
            newPos[1]= col-1;
        }
        if (direction == Directions.RIGHT){
            newPos[0]=row;
            newPos[1]= col+1;
        }

        return newPos;
    }

    private int[] randomEmptyPos(){
        int[] posicion = new int[2];
        int row = Dice.randomPos(nRows);
        int col = Dice.randomPos(nCols);
        while (!posOK(row, col) || !emptyPos(row, col)){
            row = Dice.randomPos(nRows);
            col = Dice.randomPos(nCols);

        }
        posicion[0]= row;
        posicion[1]= col;
        return posicion;
    }

    private Monster putPlayer2D(Integer oldRow ,Integer oldCol , Integer row , Integer col , Player player){
        Monster output = null;
        if(this.canStepOn(row, col)){
            if(this.posOK(oldRow, oldCol)){
                Player p = this.players[oldRow][oldCol];
                
                if(p == player){
                    this.updateOldPos(oldRow, oldCol);
                    this.players[oldRow][oldCol] =null;
                }

            }
            boolean monsterPos = this.monsterPos(row, col);
            if(monsterPos){
                this.labyrinth[row][col] = COMBAT_CHAR;
                output = this.monsters[row][col] ;  
            }
            else{
                this.labyrinth[row][col] = player.getNumber();
            }
            this.players[row][col]=player;
            player.setPos(row, col);
        }
        return output;
    }


}

File: JUEGO-RUBY/irrgarten/Labyrinth.rb
require_relative 'Directions'
require_relative 'Dice'
require_relative 'Orientation'
require_relative 'Player'
require_relative 'Monster'
module Irrgarten
    class Labyrinth
        @@BLOCK_CHAR = 'X'
        @@EMPTY_CHAR = '-'
        @@COMBAT_CHAR = 'C'
        @@MONSTER_CHAR = 'M'
        @@EXIT_CHAR = 'E'
        @@ROW = 0
        @@COL = 1
        def initialize(nrows,ncols,erow,ecol)
            @n_rows = nrows
            @n_cols = ncols
            @exit_row = erow
            @exit_col = ecol
            @monsters = Array.new(@n_rows){Array.new(@n_cols)}
            @labyrinth = Array.new(@n_rows){Array.new(@n_cols,@@EMPTY_CHAR)}
            @players = Array.new(@n_rows){Array.new(@n_cols)}
            @labyrinth[erow][ecol] = @@EXIT_CHAR
        end

        def spread_players(players)
            for p in players
                pos = random_empty_pos()
                monster = put_player_2D(-1,-1,pos[@@ROW],pos[@@COL],p)
            end
        end
        def have_a_winner
            return @players[@exit_row][@exit_col] != nil

        end
        def to_s
            cadena =""
            decor = " * "* (@n_cols+2)
            cadena.concat(decor,"\n")
            for i in 0...@n_rows
                cadena.concat(" * ")
                for j in 0...@n_cols
                    cadena.concat(" ",@labyrinth[i][j]," ")
                end
                cadena.concat(" *\n")
            end
            cadena.concat(decor,"\n")
            return cadena
        end
        def add_monster(row,col,monster)
            if pos_ok(row,col) && empty_pos(row,col)
                @labyrinth[row][col] = @@MONSTER_CHAR
                @monsters[row][col] = monster
                @monsters[row][col].set_pos(row,col)
                
            end
        end
        def add_block(orientation , start_row ,start_col, lenght)
            if orientation == Orientation::VERTICAL 
                incRow = 1
                incCol = 0
            else
                incRow = 0
                incCol = 1

            end
            
            row = start_row
            col = start_col
            
            while(pos_ok(row,col) && empty_pos(row,col) && (lenght > 0))
                @labyrinth[row][col] = @@BLOCK_CHAR
                row += incRow
                col += incCol
                lenght -=1
            end

        end
        def valid_moves(row,col)
            directions = Array.new()
            if can_step_on(row+1,col)
                directions.push(Directions::DOWN)
            end
            if can_step_on(row-1,col)
                directions.push(Directions::UP)
            end
            if can_step_on(row,col+1)
                directions.push(Directions::RIGHT)
            end
            if can_step_on(row,col-1)
                directions.push(Directions::LEFT)
            end
            return directions
        end
        def put_player(direction,player)
            old_row = player.row
            old_col = player.col
            posicion = dir_2_pos(old_row, old_col, direction)
            put_player_2D(old_row, old_col, posicion[@@ROW], posicion[@@COL], player)
        end

        private

        def pos_ok(row ,col)
            return (0 <= row && row < @n_rows && 0 <= col && col < @n_cols)
        end
        def empty_pos(row,col)
            return @labyrinth[row][col] == @@EMPTY_CHAR
        end
        def monster_pos(row,col)
            return @labyrinth[row][col] == @@MONSTER_CHAR
        end
        def exit_pos(row,col)
            return @labyrinth[row][col] == @@EXIT_CHAR
        end
        def combat_pos(row,col)
            return @labyrinth[row][col] == @@COMBAT_CHAR
        end
        def can_step_on(row,col)
            return pos_ok(row,col) && (empty_pos(row,col)||monster_pos(row,col)||exit_pos(row,col))
        end
        def update_old_pos(row,col)
            if pos_ok(row,col) 
                if combat_pos(row,col)
                    @labyrinth[row][col] = @@MONSTER_CHAR
                else
                    @labyrinth[row][col] =@@EMPTY_CHAR
                end
            end
        end
        def dir_2_pos(row,col,direction)
            if direction == Directions::UP 
                return [row-1,col]
            elsif direction == Directions::DOWN
                return [row+1,col]
            elsif direction == Directions::RIGHT 
                return [row,col+1]
            else
                return [row,col-1]
            end
        end
        def random_empty_pos
            col = Dice.random_pos(@n_cols)
            row = Dice.random_pos(@n_rows)
            while (!pos_ok(row,col) || !empty_pos(row,col))
                col =Dice.random_pos(@n_cols)
                row =Dice.random_pos(@n_rows)
            end
            return [row,col]
        end
        def put_player_2D(old_row,old_col,row,col,player)
            output = nil
            if can_step_on(row,col)
                if pos_ok(old_row,old_col)
                    jugador = @players[old_row][old_col]
                    if player == jugador
                        update_old_pos(old_row, old_col)
                        player.set_pos(old_row,old_col)
                    end
                end
                if monster_pos(row,col)
                    @labyrinth[row][col]=@@COMBAT_CHAR
                    output = @monsters[row][col]
                else
                    @labyrinth[row][col] = player.number.to_s
                end   
            
                @players[row][col] = player  
                player.set_pos(row,col)
            end
            return output
        end
    
    end
end
